macOS Bitwarden-1.16.0-beta1 safari extension problem

Hi,

i'm using the Bitwarden-1.16.0-beta1 from [here](https://github.com/bitwarden/browser/issues/664#issuecomment-525021821) that contains the bundled Safari browser extension for upcomming macOS catalina.

Logging in the brwoser extension shows a "preflight response is not successful" error. The desktop app works without problems.

server log:

[2019-08-31 14:12:09][rocket::rocket][INFO] OPTIONS /api/accounts/prelogin:
[2019-08-31 14:12:09][][ERROR] No matching routes for OPTIONS /api/accounts/prelogin.
[2019-08-31 14:12:09][][WARN] Responding with 404 Not Found catcher.
[2019-08-31 14:12:09][_][INFO] Response succeeded.

--- a/src/main.rs
+++ b/src/main.rs
@@ -250,7 +250,8 @@ fn launch_rocket() {
     let rocket = rocket
         .manage(db::init_pool())
         .manage(api::start_notification_server())
-        .attach(util::AppHeaders());
+        .attach(util::AppHeaders())
+        .attach(util::CORS());
 
     // Launch and print error if there is one
     // The launch will restore the original logging level

--- a/src/util.rs
+++ b/src/util.rs
@@ -4,6 +4,8 @@
 use rocket::fairing::{Fairing, Info, Kind};
 use rocket::response::{self, Responder};
 use rocket::{Request, Response};
+use rocket::http::{Header, ContentType, Method};
+use std::io::Cursor;
 
 pub struct AppHeaders();
 
@@ -27,6 +29,39 @@ impl Fairing for AppHeaders {
         // Disable cache unless otherwise specified
         if !res.headers().contains("cache-control") {
             res.set_raw_header("Cache-Control", "no-cache, no-store, max-age=0");
+        }
+    }
+}
+
+
+pub struct CORS();
+
+impl Fairing for CORS {
+    fn info(&self) -> Info {
+        Info {
+            name: "Add CORS headers to requests",
+            kind: Kind::Response
+        }
+    }
+
+    fn on_response(&self, request: &Request, response: &mut Response) {
+        // We need to explictly get the Origin header for Access-Control-Allow-Origin
+        let origin = match request.headers().get_one("Origin") {
+            Some(h) => h.to_string(),
+            _ => "".to_string(),
+        };
+
+        if request.method() == Method::Options || response.content_type() == Some(ContentType::JSON) {
+            // Requests with credentials need explicit values since they do not allow wildcards.
+            response.set_header(Header::new("Access-Control-Allow-Origin", origin));
+            response.set_header(Header::new("Access-Control-Allow-Methods", "GET, HEAD, POST, PUT, DELETE, CONNECT, OPTIONS, TRACE, PATCH"));
+            response.set_header(Header::new("Access-Control-Allow-Headers", "*, Authorization"));
+            response.set_header(Header::new("Access-Control-Allow-Credentials", "true"));
+        }
+
+        if request.method() == Method::Options {
+            response.set_header(ContentType::Plain);
+            response.set_sized_body(Cursor::new(""));
         }
     }
 }

--- a/src/util.rs
+++ b/src/util.rs
@@ -4,7 +4,7 @@
 use rocket::fairing::{Fairing, Info, Kind};
 use rocket::response::{self, Responder};
 use rocket::{Request, Response};
-use rocket::http::{Header, ContentType, Method};
+use rocket::http::{Header, HeaderMap, ContentType, Method, Status};
 use std::io::Cursor;
 
 pub struct AppHeaders();
@@ -33,8 +33,16 @@ impl Fairing for AppHeaders {
     }
 }
 
-
 pub struct CORS();
+
+impl CORS {
+    fn get_header(headers: &HeaderMap, name: &str) -> String {
+        match headers.get_one(name) {
+            Some(h) => h.to_string(),
+            _ => "".to_string(),
+        }
+    }
+}
 
 impl Fairing for CORS {
     fn info(&self) -> Info {
@@ -45,21 +53,25 @@ impl Fairing for CORS {
     }
 
     fn on_response(&self, request: &Request, response: &mut Response) {
-        // We need to explictly get the Origin header for Access-Control-Allow-Origin
-        let origin = match request.headers().get_one("Origin") {
-            Some(h) => h.to_string(),
-            _ => "".to_string(),
-        };
+        let req_headers = request.headers();
+
+        // We need to explicitly get the Origin header for Access-Control-Allow-Origin
+        let req_allow_origin = CORS::get_header(&req_headers, "Origin");
+
+        let req_allow_headers = CORS::get_header(&req_headers, "Access-Control-Request-Headers");
+
+        let req_allow_methods =CORS::get_header(&req_headers,"Access-Control-Request-Methods");
 
         if request.method() == Method::Options || response.content_type() == Some(ContentType::JSON) {
             // Requests with credentials need explicit values since they do not allow wildcards.
-            response.set_header(Header::new("Access-Control-Allow-Origin", origin));
-            response.set_header(Header::new("Access-Control-Allow-Methods", "GET, HEAD, POST, PUT, DELETE, CONNECT, OPTIONS, TRACE, PATCH"));
-            response.set_header(Header::new("Access-Control-Allow-Headers", "*, Authorization"));
+            response.set_header(Header::new("Access-Control-Allow-Origin", req_allow_origin));
+            response.set_header(Header::new("Access-Control-Allow-Methods", req_allow_methods));
+            response.set_header(Header::new("Access-Control-Allow-Headers", req_allow_headers));
             response.set_header(Header::new("Access-Control-Allow-Credentials", "true"));
         }
 
         if request.method() == Method::Options {
+            response.set_status(Status::Ok);
             response.set_header(ContentType::Plain);
             response.set_sized_body(Cursor::new(""));
         }

